wpiutil/src/main/native/include/wpi/struct/SchemaParser.h

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <stdint.h>
+
+#include <string>
+#include <string_view>
+#include <vector>
+#include <utility>
+
+namespace wpi::rawstruct::parser {
+
+struct Token {
+  enum Kind {
+    kUnknown,
+    kInteger,
+    kIdentifier,
+    kLeftBracket,
+    kRightBracket,
+    kLeftBrace,
+    kRightBrace,
+    kColon,
+    kSemicolon,
+    kComma,
+    kEquals,
+    kEndOfInput,
+  };
+
+  Token() = default;
+  Token(Kind kind, std::string_view text) : kind{kind}, text{text} {}
+
+  bool Is(Kind k) const { return kind == k; }
+
+  Kind kind = kUnknown;
+  std::string_view text;
+};
+
+std::string_view ToString(Token::Kind kind);
+
+class Lexer {
+ public:
+  explicit Lexer(std::string_view in) : m_in{in} {}
+
+  Token Scan();
+
+  size_t GetPosition() const { return m_tokenStart; }
+
+ private:
+  Token ScanInteger();
+  Token ScanIdentifier();
+
+  void Get() {
+    if (m_pos < m_in.size()) {
+      [[likely]] m_current = m_in[m_pos];
+    } else {
+      m_current = -1;
+    }
+    ++m_pos;
+  }
+
+  void Unget() {
+    if (m_pos > 0) {
+      [[likely]] m_current = m_in[--m_pos];
```

Comment:
```suggestion
      [[likely]] m_pos--;
      if (m_pos < m_in.size()) {
        [[likely]] m_current = m_in[m_pos];
      } else {
        m_current = -1;
      }
```
Not checking the upper bound caused the Win64Debug CI (the only one to actually run the new tests) to fail, since it would try to access an illegal index when calling `Unget()` after `Get()`-ing EOF, e.g. when scanning an identifier or integer at the end of the input.

Reply:
Thanks!

Reply:
Using GitHub's search through the logs didn't show any results for `StructParserTest` or `DynamicStructTest`, so I assumed that those tests weren't being run. Also strange that it didn't trigger ASAN...

Reply:
I think it doesn’t trigger ASAN because it’s always being given a null terminated string, so reading 1 past the end is fine from a memory perspective.  These tests aren’t platform dependent so any Gradle build will run them.

Reply:
Makes sense, also turns out GitHub's log search is a little glitchy- It seems you open the step's output for the search to work (though if you opened it earlier it will still work).